### DISCUSSION TITLE
Attribute based on other attribute

### DISCUSSION THREAD

**Discussion by romanzdk** [pydantic/pydantic#9749] [Question]
Hi I would like to define a pydantic BaseModel class (actually a pydantic settings) where the user would submit attribute e.g. START_DATE and the other attribute (e.g. END_DATE) would get computed automatically based on the START_DATE attribute. I.e. something like this:

```python
class Settings(pydantic_settings.BaseSettings):
    START_DATE: str
    END_DATE: datetime.date # = START_DATE + datetime.timedelta(days=7)
```

So it can be then called just with START_DATE (no need to explicitly set the END_DATE.

How to accomplish this please?

**Comment by Viicos** [comment]
Two solutions for your use case:

- Using a [`computed_field`](https://docs.pydantic.dev/latest/concepts/fields/#the-computed_field-decorator):
```python
from datetime import date, timedelta
from pydantic import BaseModel, computed_field


class Settings(BaseModel):
    START_DATE: date

    @property
    @computed_field
    def END_DATE(self) -> date:
        return  self.START_DATE + timedelta(days=7)
```

- Using a validator if you still want to keep the ability for the user to provide a value for `END_DATE`:

```python
from datetime import date, timedelta
from typing import Annotated, Any

from pydantic import BaseModel, Field, ValidationError, ValidationInfo, ValidatorFunctionWrapHandler, WrapValidator


def from_start_date(v: Any, handler: ValidatorFunctionWrapHandler, info: ValidationInfo) -> date:
    try:
        return handler(v)
    except ValidationError:
        return info.data['START_DATE'] + timedelta(days=7)


class Settings(BaseModel):
    START_DATE: date
    END_DATE: Annotated[date, WrapValidator(from_start_date), Field(validate_default=True)] = None  # type: ignore
```

The second solution is really not pretty at all, but is the only way I found to make it work with type checking (i.e. `Settings(START_DATE="2024-01-01")` type checks correctly, and `Settings(...).END_DATE` is of type `date`). The way it works is if no value is provided by the user, the `from_start_date` validator will be called with the default value of `None` (because we set `validate_default`), will fail to call `handler(v=None)`, and will thus default to get the value from `START_DATE`.

@sydney-runkle, any ideas how we could facilitate such a pattern? Setting a value depending from the others if not provided is a quite common pattern :thinking:

**Reply by sydney-runkle** [reply]
@Viicos,

Great point, I think this issue addresses that concern: https://github.com/pydantic/pydantic/issues/9789

**Reply by Viicos** [reply]
Thanks, seems similar indeed

**Comment by orfisko** [comment]
This is also an option if you want to be able to set an end date yourself and overwrite the compute, I use this pattern quite a lot in my application

```python
from datetime import date, timedelta, datetime
from typing import Optional

from pydantic import BaseModel, computed_field, PrivateAttr


class Settings(BaseModel):
    START_DATE: date
    _END_DATE: date = PrivateAttr(None)

    def __init__(self, END_DATE: Optional[date] = None, **data):
        super().__init__(**data)
        if END_DATE is not None:
            self._END_DATE = END_DATE


    @computed_field
    @property
    def END_DATE(self) -> date:
        if self._END_DATE is None:
            return  self.START_DATE + timedelta(days=7)
        return self._END_DATE

    @END_DATE.setter
    def END_DATE(self, user_date: date):
        assert isinstance(user_date, date)
        self._END_DATE = user_date
        
settings1 = Settings(START_DATE=datetime.strptime("01/01/2022", "%m/%d/%Y").date())
print(settings1.END_DATE)
# 2022-01-08
settings2 = Settings(START_DATE=datetime.strptime("01/01/2022", "%m/%d/%Y").date(),
                     END_DATE= datetime.strptime("03/20/2022", "%m/%d/%Y").date())

print(settings2.END_DATE)
# 2022-03-20


```
